src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp

```diff
@@ -289,138 +289,98 @@ void ShenandoahBarrierSetAssembler::load_reference_barrier_not_null(MacroAssembl
 
   Address gc_state(thread, in_bytes(ShenandoahThreadLocalData::gc_state_offset()));
   __ testb(gc_state, ShenandoahHeap::HAS_FORWARDED);
-  __ jccb(Assembler::zero, done);
-
-  // Use rsi for src address
-  const Register src_addr = rsi;
-  // Setup address parameter first, if it does not clobber oop in dst
-  bool need_addr_setup = (src_addr != dst);
-
-  if (need_addr_setup) {
-    __ push(src_addr);
-    __ lea(src_addr, src);
-
-    if (dst != rax) {
-      // Move obj into rax and save rax
-      __ push(rax);
-      __ movptr(rax, dst);
+  __ jcc(Assembler::zero, not_fwded);
+
+  Register tmp1 = noreg;
+  if (!weak) {
+    // Test for object in cset
+    // Allocate tmp-reg.
+    for (int i = 0; i < 8 && !tmp1->is_valid(); i++) {
+      Register r = as_Register(i);
+      if (r != rsp && r != rbp && r != dst && r != src.base() && r != src.index()) {
+        tmp1 = r;
+      }
     }
-  } else {
-    // dst == rsi
-    __ push(rax);
-    __ movptr(rax, dst);
-
-    // we can clobber it, since it is outgoing register
-    __ lea(src_addr, src);
+    __ push(tmp1);
+    assert(tmp1 != dst, "");
+    assert(tmp1 != src.base(), "");
+    assert(tmp1 != src.index(), "");
```

Comment:
Is this just `assert_different_registers(tmp1, dst, src.base(), src.index())`, or am I missing something?

Reply:
Not quite. dst can still legitimately be src.base(). Also, assert_different_registers() is somewhat unpractical because it fails in the wrong place. Anyhow, I rewrote it to be more succinct.